### PR TITLE
COMP: itkFFTWFFTImageFilterInitFactory.cxx is required for ITK_USE_CUFFTW

### PR DESCRIPTION
**Pull request by darrent1974**
Reinstate itkFFTWFFTImageFilterInitFactory.cxx to ITKFFT_SRCS generally. Call to FFTWFFTImageFilterInitFactory is required for the python wrapped FFT module in cases where ITK_USE_CUFFTW is defined.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Filtering/FFT/src/CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,10 @@
 set(ITKFFT_SRCS
 itkComplexToComplexFFTImageFilter.cxx
-itkVnlFFTImageFilterInitFactory.cxx)
+itkVnlFFTImageFilterInitFactory.cxx
+itkFFTWFFTImageFilterInitFactory.cxx)
```
Comment (tbirdso) [inline Modules/Filtering/FFT/src/CMakeLists.txt:4]: For best practice, FFTW sources should be included in the build only if FFTW (or CUFFTW) is explicitly requested. This line should be removed and the conditional block below should be revisited.

File: Modules/Filtering/FFT/src/CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 
 if(ITK_USE_FFTWF OR ITK_USE_FFTWD AND NOT ITK_USE_CUFFTW)
   list(APPEND ITKFFT_SRCS itkFFTWGlobalConfiguration.cxx)
-  list(APPEND ITKFFT_SRCS itkFFTWFFTImageFilterInitFactory.cxx)
```
Comment (tbirdso) [inline Modules/Filtering/FFT/src/CMakeLists.txt:7]: We can preserve conditions on adding `itkFFTWFFTImageFilterInitFactory.cxx` to sources while fixing CUFFTW builds with something like this:

```sh
if(ITK_USE_FFTWF OR ITK_USE_FFTWD)
  list(APPEND ITKFFT_SRCS itkFFTWFFTImageFilterInitFactory.cxx)
  if(NOT ITK_USE_CUFFTW)
    list(APPEND ITKFFT_SRCS itkFFTWGlobalConfiguration.cxx)
  endif()
endif()
```

**Review by tbirdso — CHANGES_REQUESTED** [review]
@darrent1974 Thank you for pinning down this issue! This may resolve https://github.com/InsightSoftwareConsortium/ITK/issues/3744.

In practice FFTW sources should be included only if the FFTW library is provided to ITK. I've suggested a small change to adhere to that practice while addressing CUFFTW considerations.

**Comment by tbirdso** [comment]
Hi @darrent1974 , I am still seeing ITKFFT Python import failures when I run a local build with these changes applied. Can you confirm that the fixes here resolve issues on your end? And, could you please provide additional details about your platform, including OS, ITK version, Python version, and NVIDIA GPU Computing Toolkit version? Thanks.

I've logged my observations in https://github.com/InsightSoftwareConsortium/ITK/issues/4028, please feel free to add discussion to that issue regarding the faulty behavior you've observed.

**Comment by darrent1974** [comment]
Hi @tbirdso, it's all ok on my end. It's actually running in a live system as of last week.

I suspect one major difference may be that you are enabling ITK_USE_GPU? I'm not, currently I'm only interested in using ITK and cuFFT for FFT's. However, I do recall having some build issues when I was playing around with enabling ITK_USE_GPU. This may become an issue for us further down the road when I transition to native ITK CUDA implementations. However, at this point, our platform doesn't require us to enable ITK_USE_GPU.

Details of the platform are:

Linux containerised environment:

Base image: nvidia/cuda:12.1.0-devel-ubuntu22.04
CMake 3.24.1
Python 3.11
CUDA 12.1.0 
ITK 5.3.0

Relevant ITK Configuration

    cmake \
      -DCMAKE_BUILD_TYPE:STRING=${build_type} \
      -DITK_SOURCE_DIR:PATH=${source_path} \
      -DITK_BINARY_DIR:PATH=${build_path} \
      -DBUILD_TESTING:BOOL=OFF \
      -DBUILD_DOCUMENTATION:BOOL=OFF \
      -DCMAKE_CXX_STANDARD_REQUIRED:BOOL=ON \
      -DCMAKE_CXX_STANDARD=17 \
      -DPython3_EXECUTABLE:FILEPATH=${Python3_EXECUTABLE} \
      -DPython3_INCLUDE_DIR:PATH=${Python3_INCLUDE_DIR} \
      -DCMAKE_CXX_COMPILER_TARGET:STRING="$(uname -p)-linux-gnu" \
      -DCMAKE_CXX_FLAGS:STRING="$compile_flags" \
      -DCMAKE_C_FLAGS:STRING="$compile_flags" \
      -DCMAKE_BUILD_TYPE:STRING="${build_type}" \
      -DWRAP_ITK_INSTALL_COMPONENT_IDENTIFIER:STRING=PythonWheel \
      -DWRAP_ITK_INSTALL_COMPONENT_PER_MODULE:BOOL=ON \
      -DPY_SITE_PACKAGES_PATH:PATH="." \
      -DITK_USE_64BITS_IDS:BOOL=ON \
      -DITK_LEGACY_SILENT:BOOL=ON \
      -DITK_WRAP_PYTHON:BOOL=ON \
      -DITK_WRAP_DOC:BOOL=OFF \
      -DModule_ITKTBB:BOOL=ON \
      -DTBB_DIR:PATH=${tbb_dir} \
      -DModule_ITKImageFusion:BOOL=ON \
      -DITK_USE_FFTWD:BOOL=ON \
      -DITK_USE_FFTWF:BOOL=ON \
      -DITK_USE_SYSTEM_FFTW:BOOL=ON \
      -DITK_USE_CUFFTW:BOOL=ON \
      -DCUFFTW_INCLUDE_PATH=/usr/local/cuda/include \
      -DCUFFTW_LIB=/usr/local/cuda/lib64/libcufftw.so \
      -DCUFFT_LIB=/usr/local/cuda/lib64/libcufft.so \
      -DITK_USE_SYSTEM_HDF5:BOOL=ON \
      -DITK_USE_GPU:BOOL=OFF \
      -DOPENCL_LIBRARIES=/usr/local/cuda/lib64/libOpenCL.so \
      -DITK_WRAP_float:BOOL=ON \
      -DITK_WRAP_double:BOOL=ON \
      -DITK_WRAP_complex_double:BOOL=ON \
      -DITK_WRAP_unsigned_short:BOOL=ON \
      -DITK_WRAP_unsigned_long_long:BOOL=ON \
      -DITK_WRAP_IMAGE_DIMS:STRING="1;2;3;4" \
      -DITK_WRAP_VECTOR_COMPONENTS:STRING="1;2;3;4;5" \
      ${CMAKE_OPTIONS} \
      -G Ninja \
      ${source_path}

**Review by tbirdso — APPROVED** [review]

**Pull request merged**